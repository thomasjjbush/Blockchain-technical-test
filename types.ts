import { AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { Json } from 'enzyme-to-json';
import HtmlWebPackPlugin from 'html-webpack-plugin';

export type Action = ThunkAction<Promise<AnyAction> | AnyAction, Store, unknown, AnyAction>;

export interface Address {
    address?: string;
    error?: boolean;
    final_balance?: number;
    hash160?: string;
    n_tx?: number;
    total_received?: number;
    total_sent?: number;
    tx_per_page: number;
    txs?: Txs[];
}

export enum AddressActions {
    ERROR = 'address/ERROR',
    LOADED = 'address/LOADED',
}

export type Dispatch = ThunkDispatch<Store, void, AnyAction>;

export type FetchParams = Record<string, boolean | number | string>;

export interface Home {
    error?: boolean;
    latest_transactions: Txs[];
}

export enum HomeActions {
    ERROR = 'home/ERROR',
    LATEST_TRANSACTION = 'home/LATEST_TRANSACTION',
}

export interface MultiPageButtonsProps {
    currentPage: number;
    numberOfButtons: number;
    numberOfPages: number;
    skip?: number;
    to: (i: number) => Partial<Location>;
}

export interface SearchInputProps {
    onChange: (text: string) => void;
    placeholder?: string;
    value: string;
}

export interface SerializerMap extends Json {
    node: {
        type: {
            componentStyle?: {
                baseStyle?: {
                    rules: (string | ((props: StyledProps) => string))[];
                };
                rules: (string | ((props: StyledProps) => string))[];
            };
        };
    };
}

export interface SocketOptions<T> {
    onClose: string;
    onError: T;
    onMessage: T;
    onOpen: string;
}

export interface Store {
    address: Address;
    home: Home;
}

export type StyledProps<Props = {}> = { theme: Theme } & Props;

export interface Theme {
    colors: {
        brand: string;
        offset: string;
    };
}

export interface TransactionProps extends Partial<Txs> {
    search?: string;
}

export interface Txs {
    balance: number;
    fee: number;
    hash: string;
    result: number;
    size: number;
}

export interface WebpackConfig {
    devServer: { historyApiFallback: boolean; hot: boolean; open: boolean };
    entry: string[];
    mode: string;
    module: {
        rules: WebpackRule[];
    };
    output: { path?: string; publicPath: string };
    plugins: HtmlWebPackPlugin[];
    resolve: { extensions: string[] };
}

interface WebpackRule {
    exclude?: (RegExp | string)[];
    test: RegExp;
    use: {
        loader: string;
        options?: { plugins: string[] };
    };
}
