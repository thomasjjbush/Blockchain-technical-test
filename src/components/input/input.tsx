import React, { ChangeEvent, FC, ReactElement } from 'react';
import styled from 'styled-components';
import { SearchInputProps as Props } from '../../../types';

const Container = styled.span`
    border: solid 2px ${(props): string => props.theme.colors.offset};
    border-radius: 5px;
    overflow: hidden;
`;

const Button = styled.button`
    border: none;
    height: 100%;
    padding: 0 10px;
`;

const Input = styled.input`
    border: none;
    height: 100%;
    padding: 0 10px;
`;

export const SearchInput: FC<Props> = ({ onChange, value, placeholder }: Props): ReactElement => {
    return (
        <Container>
            <Input
                data-test-id="search"
                onChange={(e: ChangeEvent<HTMLInputElement>) => onChange(e.currentTarget.value)}
                placeholder={placeholder}
                value={value}
                type="text"
            />
            <Button data-test-id="clear" disabled={!value} onClick={() => onChange('')}>
                x
            </Button>
        </Container>
    );
};
