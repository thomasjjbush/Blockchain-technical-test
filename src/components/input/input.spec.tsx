import React from 'react';
import { shallow } from 'enzyme';
import { SearchInput } from './input';
import { SearchInputProps as Props } from '../../../types';

describe('SearchInput', () => {
    let props: Props;

    beforeEach(() => {
        props = {
            onChange: jest.fn(),
            value: '',
        };
    });

    it('should match snapshot', () => {
        const wrapper = shallow(<SearchInput {...props} placeholder="placeholder" value="value" />);
        expect(wrapper).toMatchSnapshot();
    });

    it('should invoke onChange cb on input change', () => {
        const wrapper = shallow(<SearchInput {...props} />);
        wrapper.find({ 'data-test-id': 'search' }).prop('onChange')({ currentTarget: { value: 'howdy' } });

        expect(props.onChange).toHaveBeenCalledTimes(1);
        expect(props.onChange).toHaveBeenCalledWith('howdy');
    });

    it('should disable clear button if value resolves to false', () => {
        const wrapper = shallow(<SearchInput {...props} />);
        expect(wrapper.find({ 'data-test-id': 'clear' }).prop('disabled')).toBe(true);

        wrapper.setProps({ value: 'something else' });
        expect(wrapper.find({ 'data-test-id': 'clear' }).prop('disabled')).toBe(false);
    });

    it('should invoke onChange cb with empty string on clear click', () => {
        const wrapper = shallow(<SearchInput {...props} />);
        wrapper.find({ 'data-test-id': 'clear' }).simulate('click');
        expect(props.onChange).toHaveBeenCalledTimes(1);
        expect(props.onChange).toHaveBeenCalledWith('');
    });
});
