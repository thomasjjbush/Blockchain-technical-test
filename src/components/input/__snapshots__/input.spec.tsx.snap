// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput should match snapshot 1`] = `
<styled.span
  styles="
    border: solid 2px theme.colors.offset;
    border-radius: 5px;
    overflow: hidden;
"
>
  <styled.input
    data-test-id="search"
    onChange={[Function]}
    placeholder="placeholder"
    styles="
    border: none;
    height: 100%;
    padding: 0 10px;
"
    type="text"
    value="value"
  />
  <styled.button
    data-test-id="clear"
    disabled={false}
    onClick={[Function]}
    styles="
    border: none;
    height: 100%;
    padding: 0 10px;
"
  >
    x
  </styled.button>
</styled.span>
`;
